File: src/test/java/com/bigdata/leetcode/C146LRUCache.java
package com.bigdata.leetcode;

import org.junit.Test;
import sun.misc.LRUCache;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * @ClassName C146LRUCache
 * @Description TODO
 * @Author HuHao（huhao1@cmcm.com）
 * @Date 2020/8/13 16:58
 * @Version 1.0
 * 运用你所掌握的数据结构，设计和实现一个  LRU (最近最少使用) 缓存机制。它应该支持以下操作： 获取数据 get 和 写入数据 put 。
 *
 * 获取数据 get(key) - 如果关键字 (key) 存在于缓存中，则获取关键字的值（总是正数），否则返回 -1。
 * 写入数据 put(key, value) - 如果关键字已经存在，则变更其数据值；如果关键字不存在，则插入该组「关键字/值」。
 * 当缓存容量达到上限时，它应该在写入新数据之前删除最久未使用的数据值，从而为新的数据值留出空间。
 *
 * 进阶:
 *
 * 你是否可以在 O(1) 时间复杂度内完成这两种操作？
 *
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/lru-cache
 *
 * LRUCache cache = new LRUCache( 2/ ); //缓存容量
* cache.put(1,1);
* cache.put(2,2);
* cache.get(1);       // 返回  1
* cache.put(3,3);    // 该操作会使得关键字 2 作废
* cache.get(2);       // 返回 -1 (未找到)
* cache.put(4,4);    // 该操作会使得关键字 1 作废
* cache.get(1);       // 返回 -1 (未找到)
* cache.get(3);       // 返回  3
* cache.get(4);       // 返回  4

 **/


/**
 * 双向链表 + hash
 * 时间复杂度 o(1)
 * 空间复杂度 o(capacity+1)
 */
class LRUCache1 {

    private int capacity = 2;
    private int size = 0;
    private Map<Object, LinkedNode> cache = null;
    private LinkedNode head = null;
    private LinkedNode tail = null;

    public LRUCache1() {
        this.cache = new HashMap<>();
        this.head = new LinkedNode(null, null);
        this.tail = new LinkedNode(null, null);
        this.head.next = this.tail;
        this.tail.prev = this.head;
    }

    public LRUCache1(int capacity) {
        this();
        this.capacity = capacity;
    }

    public Object get(Object key) {
        Object result = null;
        if (cache.containsKey(key)) {
            LinkedNode node = cache.get(key);
            result = node.value;
            moveToHead(node);
        }
        return result;
    }

    public void put(Object key, Object value) {
        if (cache.containsKey(key)) {
            LinkedNode node = cache.get(key);
            node.value = value;
            moveToHead(node);
        } else {
            LinkedNode node = new LinkedNode(key, value);
            cache.put(key,node);
            addToHead(node);
            if (size > capacity) {
                cache.remove(removeTail());
            }
        }
    }

    private void moveToHead(LinkedNode node) {
        remove(node);
        addToHead(node);
    }

    private void addToHead(LinkedNode node) {
        LinkedNode second = head.next;
        node.prev = head;
        node.next = second;
        second.prev = node;
        head.next = node;
        size ++;
    }

    private Object removeTail() {
        LinkedNode prev = tail.prev;
        remove(prev);
        return prev.key;
    }

    private void remove(LinkedNode node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
        size --;
    }

    class LinkedNode {
        Object key = null;
        Object value = null;
        LinkedNode prev = null;
        LinkedNode next = null;

        public LinkedNode(Object key, Object value) {
            this.key = key;
            this.value = value;
        }
    }
}

/**
 * 基于 LinkedHashMap 自带容量约束，复写 过期函数 removeEldestEntry
 */
class LRUCache2 extends LinkedHashMap<Object,Object> {

    int capacity = 2;

    public LRUCache2(int capacity) {
        // accessOrder 维持顺序
        super(capacity,0.75F,true);
        this.capacity = capacity;
    }

    @Override
    public Object get(Object key) {
        return super.getOrDefault(key,null);
    }

    @Override
    public Object put(Object key, Object value) {
        return super.put(key, value);
    }

    @Override
    protected boolean removeEldestEntry(Map.Entry<Object, Object> eldest) {
        return size()>capacity;
    }
}


public class C146LRUCache {

    @Test
    public void solve1Test(){
        LRUCache1 lruCache = new LRUCache1(2);
        lruCache.put(1, 1);
        lruCache.put(2, 2);
        System.out.println(lruCache.get(1));// 刚满，可以取出
        lruCache.put(3, 3); // 挤出一个，由于2最近使用最少，所以挤出2
        System.out.println(lruCache.get(2)); // 取不到 null
        lruCache.put(4, 4); // 挤出一个，3 是刚存进去的，所以1被挤出
        System.out.println(lruCache.get(1)); // 取不到 null
        System.out.println(lruCache.get(3)); // 取出3 （且3被前置）
        System.out.println(lruCache.get(4)); // 取出4 （且4被前置）
    }

    @Test
    public void solve2Test(){
        LRUCache2 lruCache = new LRUCache2(2);
        lruCache.put(1, 1);
        lruCache.put(2, 2);
        System.out.println(lruCache.get(1));// 刚满，可以取出
        lruCache.put(3, 3); // 挤出一个，由于2最近使用最少，所以挤出2
        System.out.println(lruCache.get(2)); // 取不到 null
        lruCache.put(4, 4); // 挤出一个，3 是刚存进去的，所以1被挤出
        System.out.println(lruCache.get(1)); // 取不到 null
        System.out.println(lruCache.get(3)); // 取出3 （且3被前置）
        System.out.println(lruCache.get(4)); // 取出4 （且4被前置）
    }


}

File: python2/leetcode/C146LRUCache.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Author: HuHao <whohow20094702@163.com>
Date: 2020/8/13 11:59
Info:

"""

import os, sys

# 双向链表 + hash
# 时间复杂度 o(1)
# 空间复杂度 o(capacity+1)
class LRUCache:
    # 双向链表存储维护使用频率
    class LinkedNode:
        def __init__(self,key=None,value=None):
            self.key = key
            self.value = value
            self.prev = None
            self.next = None

    def __init__(self,capacity=2):
        self.capacity = capacity
        self.size = 0 # 当前容量
        self.cache = dict() # hash 存储
        self.head = LRUCache.LinkedNode() # 链头(标记位 实际不存数据)
        self.tail = LRUCache.LinkedNode() # 链尾(标记位 实际不存数据)
        self.head.next = self.tail # 链头绑定链尾
        self.tail.prev = self.head # 链尾绑定链头

    def get(self,key):
        result = None
        if key in self.cache:
            node = self.cache[key]
            result = node.value
            self._move_to_head(node) # 使用过的往链头移动
        return result

    def put(self,key,value):
        if key in self.cache:
            node = self.cache[key]
            node.value = value
            self._move_to_head(node) # 刚刷新过的往链头移动
        else:
            node = LRUCache.LinkedNode(key, value)
            self._add_to_head(node) # 刚添加的往链头移动
            self.cache[key] = node
            if self.size > self.capacity:
                key = self._remove_tail() # 容量超限的，从尾部开始清除
                self.cache.pop(key) # 注意同步删除数据

    def _remove_node(self,node):
        node.next.prev = node.prev # 下个节点的前，指向当前节点的前
        node.prev.next = node.next # 上个节点的后，指向当前节点的后
        self.size -= 1

    def _remove_tail(self):
        node = self.tail.prev # 名义上是删除 tail，其实是删除倒数第二个
        self._remove_node(node) # 执行删除
        return node.key # 抛出 key ，供hash删除

    def _add_to_head(self,node):
        # 名义上是添加到头，实质是添加到第二位
        second = self.head.next # 先标记旧二
        node.prev = self.head # 新二向前绑定头
        node.next = second # 新二想后绑定旧二
        self.head.next = node # 头向后绑定新二
        second.prev = node # 旧二向前绑定新二
        self.size += 1

    def _move_to_head(self,node):
        self._remove_node(node) # 先删除
        self._add_to_head(node) # 然后重新添加到头

# 方案2 python3/C146LRUCache.py 继承 OrderDict

if __name__ == "__main__":
    lru_cache = LRUCache(2)
    lru_cache.put(1,1)
    lru_cache.put(2,2)
    print(lru_cache.get(1)) # 刚满，可以取出
    lru_cache.put(3,3) # 挤出一个，由于2最近使用最少，所以挤出2
    print(lru_cache.get(2)) # 取不到 -1
    lru_cache.put(4,4) # 挤出一个，3 是刚存进去的，所以1被挤出
    print(lru_cache.get(1)) # 取不到 -1
    print(lru_cache.get(3)) # 取出3 （且3被前置）
    print(lru_cache.get(4)) # 取出4 （且4被前置）
